### ISSUE TITLE
Generating PDF Output

### ISSUE THREAD

**Issue by jpfairbanks** [tlienart/Franklin.jl#611]
I really like writing in Franklin, is there a way to use all the features of features of franklin's MD+Latex syntax such as commands and environments, but get out something that I could pass to Latex or Pandoc flavored markdown instead of rendering to HTML? I would use Franklin for all my document preparation if I could get Latex or Pandoc out of it.

**Comment by tlienart** [comment]
Right so that's not easy.

A path to it would be to exploit pandoc's HTML -> LaTeX conversion, do a "fixing" pass on the produced LaTeX and then compile that to PDF.

If you have very simple markup on a single page this would "just work", here are some issues that would take some effort to fix (could be done in that 'fixing' pass)

* maths, I'm pretty sure pandoc wouldn't capture the Katex properly but it might be reasonably ok to do it ourselves
* code, that would be fine but the highlighting would need to be added via the proper latex package
* ordering of pages (I guess could be specified manually) + general knitting of pages (eg do you keep the section levels?)
* table of content(s)
* custom styling (div blocks) à map would need to be manually provided from the custom CSS to custom LaTeX

It's probably worth a try and increment the stuff that's supported slowly. Suggestions with respect to the above points welcome; or more generally thoughts of how things could look like (eg single page, multi page?)

Note: in the meantime I suggested elsewhere that using Firefox to print PDF works quite well and is immediate; if you use bootstrap CSS there's even specific stuff you can do to make that PDF look exactly right, but maybe it's less elegant.

**Comment by tlienart** [comment]
Ok so I gave it a shot, here are things that seem to work ok and things that will need help. 

**works ok**: basic text formatting, headers, links, inline code, non-highlighted code, paragraphs, rule, local images
**needs work**:
- (already mentioned) maths, div blocks, code highlighting
- tables don't seem to get appropriately rendered
- images with URLs
- remove parts attributed to head.html/foot.html

A positive note is that the pipeline `pandoc + pdflatex` is fairly quick (under 1s per page from what I can tell). You can already try this out by running the following (dirty) script

```julia
for (root, _, files) in walkdir("__site")
    for f in files
        endswith(f, ".html") || continue
        src = joinpath(root, f)
        occursin("/layout/", src) && continue
        occursin("/libs/", src) && continue
        occursin("/assets/", src) && continue
        occursin("/tag/", src) && continue
        dst = splitext(src)[1] * ".tex"
        try
            open("/tmp/junk", "w") do out
                redirect_stdout(out) do
                    redirect_stderr(out) do
                        run(`pandoc --standalone $src --output $dst`)
                        run(`pdflatex -interaction=nonstopmode -quiet $dst`)
                    end
                end
            end
        catch
        end
    end
end
```

for my test websites this took 2.5 sec for 4 pages.

**Comment by epatters** [comment]
Thanks for giving this a shot and posting the experimental script.

It makes sense to me that, even with pandoc's magic, it would be hard to convert from the generated HTML to other formats. Not knowing the details of Franklin's rendering pipeline, I don't know whether this would be feasible, but I imagine it would work better to hook in at an intermediate stage of the pipeline, after the macros have been expanded and all Julia code executed, but before the HTML is generated. So, you would pass Markdown to pandoc instead of HTML. Pandoc supports lots of Markdown extensions and generally handles Markdown well. Would something like that be feasible?

**Comment by tlienart** [comment]
Yes this is a fair suggestion but with one major caveat: you would not be able to use hfuns (these `{{...}}`) 

If we start going that route though, Pandoc is not even required as far as I'm aware. The Base Markdown lib (used for text in Franklin) can output LaTeX and we could implement specific rules for things like code evals, maths, etc.

I'm not sure yet what would be the more doable path: going directly from markdown might be more straightforward but would require maintaining a larger code base with all the LaTeX paths and would not support hfun, going from HTML via pandoc is probably simpler but would probably require a bunch of hacks to "fix" the output.

**Comment by jpfairbanks** [comment]
I’m generally a believer that it is better to do things “upstream” so that you are working closer to the user input. Taking machine generated HTML and turning it into LaTex seems a lot harder than taking user generated markdown and using a markdown to latex engine. I’m not thinking of this as a method for printing a webpage to PDF, I’m thinking of writing academic papers in Franklin’s dialect of markdown with the intent of generating latex for the journal. In that workflow, restrictions on “you can’t use raw html mode or CSS styling, if you want to generate latex” seem totally fine. It is really the features around directly running Julia code and splicing in the output that I want to leverage.

**Comment by tlienart** [comment]
Yes this makes sense, thank you for your input!

**Comment by rikhuijzer** [comment]
I would very much like this feature too. For now, my main priority is creating a PDF from a single page. For multiple pages, I think the Documenter.jl `make.jl` would be a good direction. 

I have considered starting writing code, but think it is better to have a little bit of a discussion still in order to avoid working in a wrong direction. I came up with the following table. In this table, extra package requirements are listed inside the parentheses. Also, trivial things like bold text are omitted and it assumes that Franklin already evaluated the code blocks, its functions and inserted the `\tableinput`. 

Description | Franklin -> HTML | Pandoc -> LaTeX -> PDF | LaTeX -> PDF
--- | --- | --- | ---
Inline math | $ ... $ | $ ... $ | $ ... $ 
Display math | $$ ... $$ | $$ ... $$ | $$ ... $$
Headers | ## | ## | \subsection
Links | [...](...) | [...](...) | \url (hyperref)
Cross-references | **not supported?** | all kinds of inconsistent syntax (pandoc-crossref) | \label and \ref
Tables | Markdown table syntax | Markdown table syntax | \begin{tabular} etc.
Footnote | [^1] | [^1] | \footnote
Figures | ![...](...) | ![...](...) | \begin{figure} etc.; requires packages for SVG
Lists | - ... \n - ... | - ... \n - ... | \begin{itemize}
Table of Contents | \toc | **not supported?** | \tableofcontents
Raw HTML | ~~~ ... ~~~ | ... (raw_html) | **not supported**
CSS styling | .foo { ... } | **not supported** | **not supported**
References | \biblabel; \citep; \citet  | Bib(La)TeX @foo2020 | bib file; \citep; \citet
Tikz | various ways | via ![...](...) | \begin{tikzpicture} (Tikz)
Tikz cross-reference | **not supported** |  \begin{figure}; \label | \begin{figure}; \label
Code highlight | (highlight.js) | **possibly supported** | **not supported**

So, both options aren't perfect. My intuition is that avoiding Pandoc is the best choice, because
- it avoids the Pandoc dependency and its restrictions such as inconsistent cross-references syntax;
- can be combined with moving the Franklin syntax even more towards LaTeX for things which are hard or impossible to express in Markdown; that way Franklin can properly express things like cross-references and newcomers to Franklin with LaTeX experience can get started quickly

@tlienart what would be your view on this and what would be a good starting point to get something up and running? The above code snippet doesn't eval code, so for me it is not very useful.

**Comment by jpfairbanks** [comment]
Given that the Franklin syntax is evolving faster than Pandoc syntax and has support for \command{arg} style commands I think that a direct Franklin to Latex converter for a single page would be a great start.

**Comment by tlienart** [comment]
Yes, this is essentially doable for a subset of Franklin (effectively everything modulo the raw HTML stuff, div blocks and hfuns). I'll try to get a proof of concept running by the end of the month and we could build from there.

The current path would be:

input -> Franklin -> intermediate markdown -(A)> intermediate latex -(B)> injection of special blocks -> page latex

* (A) can use Julia's base Markdown -> Latex (this is essentially ok for things like text formatting etc)
* (B) is to re-plug things like code blocks and math blocks (which would not be treated properly by (A))

**Comment by rikhuijzer** [comment]
I don't fully understand the proposed path, but trust that you know what you are talking about. 

> I'll try to get a proof of concept running by the end of the month and we could build from there.

That would be awesome. Please don't feel forces into anything. Let me know if I can help out.

**Comment by tlienart** [comment]
thanks Rik!

To understand the workflow you have to understand how Franklin works. Ignoring file watching etc, it consists of the following steps from input markdown to output html:

* tokenization + forming of blocks (from-->to) for instance finding the span of `@@name ... @@`
* extraction of a markdown text (A) where all special blocks (e.g. `@@name ... @@`) have been removed and replaced by a filler (something like `FILL_SOMETHING`)
* transform `A` into HTML using Julia's Markdown module to go from `md` -> `html`
* fix the result by replacing the `FILL_SOMETHING` by what Franklin outputs for special blocks
* pass over the resulting HTML to resolve "moustache blocks" ( `{{...}}`) 

so the path I'm suggesting above is the same except that the last 3 steps would be

* ... from `md` -> `latex` (this will just work since Julia's Markdown module can do that)
* fix the result by replacing the `FILL_SOMETHING` by what Franklin **should output in latex** for special blocks
* return the resulting LaTeX, possibly do some post-processing to add some preamble or whatever

The somewhat hard bit is the second point; to find a latex-equivalent for all special blocks. To make this easier we'd just ignore raw html blocks, `@@...@@` blocks, mustache blocks etc. (By what people have said here, they wouldn't care too much about it anyway).

Harder still is to allow flexibility in this pass, for instance in LaTeX you typically want to add captions etc to figures, you might also want to specify positioning of the figure etc. One thing that I do not want to do is start a whole new syntax just for the markdown -> latex path. But we can worry about that once there's something simple that does the basic stuff right, I'm sure people will then chip in with suggestions and hopefully PR to make things better 😄

**Comment by tlienart** [comment]
Just a teaser...  (left is pdf output, right is html)

![Screenshot 2021-02-07 at 19 03 05](https://user-images.githubusercontent.com/10897531/107155185-4388de00-6977-11eb-9dff-9015e11b39b6.png)

ETA probably a couple of months (I'm profiting from the project to do a fair bit of refactoring of conversion code (md -> X, html -> X).

**Comment by jpfairbanks** [comment]
Oh, I'm so excited for this feature! Having Franklin able to generate latex would allow us to share content from our project website with published papers.

**Comment by rikhuijzer** [comment]
Great work, Thibaut! I find your perseverance impressive/inspiring

**Comment by epatters** [comment]
Looks great. I am stoked for this!
